### ISSUE TITLE
Add rejectOnNotFound to update

### ISSUE THREAD

**Issue by incompletude** [prisma/prisma#10142]
I found this issue that relates to this request, but it is closed (not sure why).

https://github.com/prisma/prisma/issues/5032

I think it makes sense to have rejectOnNotFound on `update` also, and I think very weird that is throws an exception instead of returning null.

**Comment by janpio** [comment]
Issues #5032 is closed as the ugly error was replaced with a more descriptive error: https://github.com/prisma/prisma/issues/5032#issuecomment-786576567 Can you confirm this is still the exception and error you are getting?

**Comment by jlalmes** [comment]
Hi @janpio - I am in agreement with this issue, `rejectOnNotFound` should support `update` and `delete` operations.

**Comment by afcode123** [comment]
Hi!

I think also that update should just return null if a record is not found. In my case it would be much more useful since the query is faster when it is ended immediately after record is found and it will not go through all records every time. 
I think this would be quite important addition. 
Currently I have a large table in where updating is quite slow due to the fact that I cannot use update in it since it fails on "An operation failed because it depends on one or more records that were required but not found. Record to update not found."
error each time. Instead I am using updateMany and it is quite inefficient.

**Comment by incompletude** [comment]
Totally, Prisma design forces you to put the update inside a try catch.

**Comment by SynergyEvolved** [comment]
Is there no recommended way to handle this error?

**Comment by Breck37** [comment]
+1 on this issue

**Comment by aya-eiya** [comment]
I don't know It is same request or issue here.

I can make sense on "update" or "delete" method will cause the error on NotFound.
but I have get same error also on "updateMany" method and that can't be expected.
I expect that It'll accept an update query even if effects nothing as is an sql update query does.

Is that a bug or something wrong with the setting of prisma environment?

**Comment by aya-eiya** [comment]
> I have get same error also on "updateMany" method and that can't be expected.

I found that the problem was not exactly in the updateMany method.

I was running multiple update queries using the $transaction method, the first of which was the updateMany method and the last was the update method query.

The error was for this update method, but with $transaction, all errors seem to appear as if they were for the first query in the array.

I will check if some Issue has been reported on this, and if it does not appear to have been there, I will report it as a new Issue.

and I'll also check if this can be regenerate on the latest version.

**Comment by Thore1954** [comment]
Agree with this. Also makes sense for `delete`. Currently you have to surround everything in a try catch which is not ideal.

**Comment by dani-mp** [comment]
Adding a param like this would be a nice addition!

**Comment by yawboakye** [comment]
👋 @janpio any update on this issue? it seems a little weird that the database itself throws no exception when an update doesn't happen but prisma would. bear in mind that making db calls is expensive. performing a fetch only so that one can safely perform an update is, in my opinion, too high a price to pay for the convenience of not dealing with an exception. if there are other non-query based work arounds i'd be happy to consider. i saw in the other thread that `updateMany` may be an option. my only objection here is that name of the function, since the `where` condition could be on a unique field.

**Comment by nihalanisumit** [comment]
Waiting for an update...

**Comment by meshuamam** [comment]
Also 👍  voting on this

**Comment by herzog0** [comment]
I also believe that Prisma should behave silently in this case, just like Postgres does. It is really bad to keep one rule in mind when doing raw SQL and another when coding with Prisma. We ship things really fast in my startup, and this incoherence has hit us a lot of times in production, because we are always thinking in terms of what would Postgres do with that instruction.

**Comment by elijaholmos** [comment]
Voicing my support to see this added. At a minimum, some sort of toggle should be supported. Preferably, an error isn't thrown at all.

**Comment by maximeburri** [comment]
Do you plan to develop this feature? I agree that it would makes sense to have it.

For those interested in it, it is still possible to use this workaround which use try/catch:

```typescript
    async function recoverFromNotFound<A>(
      promise: Promise<A>,
    ): Promise<A | null> {
      try {
        return await promise;
      } catch (e) {
        if (e instanceof Prisma.PrismaClientKnownRequestError) {
          if (e.code === 'P2025') {
            return null;
          }
        }
        throw e;
      }
    }
```

Here are the usage example based on the [this comment example](https://github.com/prisma/prisma/issues/10142#issuecomment-1282213935):

```typescript
    const updatedUser = await recoverFromNotFound(
      prisma.user.update({
        where: {
          id: userID,
        },
        data: {
          /* some changes */
        },
      }),
    );

    if (!updatedUser) {
      /* handle if update didn't happen */
    }
```

**Comment by disburden** [comment]
+1

**Comment by ErAz7** [comment]
you can use a simple workaround for this as long as it's not applied to the original api:

## Typescript: 
```ts
export const nullIfError =
    <Func extends AnyFunction>(func: Func) =>
    async (...args: Parameters<Func>) => {
        try {
            return await func(...args);
        } catch (err) {
            console.error(err);

            return null;
        }
    };

// then when using prisma api:
const updatedUser = await nullIfError(prisma.user.update)({
    where: {
        email: 'sth@somewhere.com',
    },
});
```

## Javascript:

```js
export const nullIfError =
    (func) =>
    async (...args) => {
        try {
            return await func(...args);
        } catch (err) {
            console.error(err);

            return null;
        }
    };

// then when using prisma api:
const updatedUser = await nullIfError(prisma.user.update)({
    where: {
        email: 'sth@somewhere.com',
    },
});
```

**Comment by jo-warren** [comment]
Borrowing from @maximeburri but as an extension as an alternative solution:
```
import { Prisma } from "@prisma/client";

/**
 * Unlike postgres, prisma throws an error when trying to update or delete a record that does not exist.
 * https://github.com/prisma/prisma/issues/10142
 */
export const IgnoreNotFoundError = Prisma.defineExtension({
    name: 'IgnoreNotFoundError',
    model: {
        $allModels: {
            async updateIgnoreNotFound<T, A>(
                this: T,
                args: Prisma.Exact<A, Prisma.Args<T, 'update'>>
            ): Promise<Prisma.Result<T, A, 'update'> | null> {
                try {
                    const context = Prisma.getExtensionContext(this) as any // skipcq: JS-0323. From prisma exension docs
                    return await context.update(args)
                } catch (err) {
                    if (err instanceof Prisma.PrismaClientKnownRequestError && err.code === 'P2025') {
                        return null
                    }
                    throw err
                }
            },
            async deleteIgnoreNotFound<T, A>(
                this: T,
                args: Prisma.Exact<A, Prisma.Args<T, 'delete'>>
            ): Promise<Prisma.Result<T, A, 'delete'> | null> {
                try {
                    const context = Prisma.getExtensionContext(this) as any // skipcq: JS-0323. From prisma exension docs
                    return await context.delete(args)
                } catch (err) {
                    if (err instanceof Prisma.PrismaClientKnownRequestError && err.code === 'P2025') {
                        return null
                    }
                    throw err
                }
            },
        }
    }
})
```

**Comment by aravindanve** [comment]
I think it is designed this way so that it allows for things like optimistic concurrency control.

```ts
// only update if the version is same as we saw last, or throw
await prisma.user.update({
  where: { id: user.id, version: user.version },
  data: { city: 'Berlin', version: { increment: 1 } },
})
```
See https://www.prisma.io/docs/orm/reference/prisma-client-reference#optimistic-concurrency-control-on-updates

Ideally `updateMany` could be used, and should have the same effect when using a unique key

```ts
const updatedUsers = await prisma.user.updateMany({
  where: {
    id: userID,
  },
  data: {
    /* some changes */
  },
});

if (!updatedUsers.length) {
  /* handle if update didn't happen */
}
```

**Comment by esmaeilzadeh** [comment]
please do not change the default behavior of update because we rely on it for optimistic concurrency control.

**Comment by DobromirKirovLime** [comment]
What is the state of this...? I was very surprised to find that many people that are complaining about this, it is very strange behaviour imo as well...

**Comment by yoniabitbol** [comment]
Up! This would be very helpful to have.

**Comment by drtz** [comment]
To be consistent with the rest of the API, all single-document operations should throw if they fail to operate on that single document. Adding a new method would just make things more confusing, and there are already clean ways to handle the scenarios mentioned above:

- Use `updateMany()` if you don't care how many documents are updated.
- In the unusual situation where you need to update exactly 0 or 1 documents, then use `update()` surrounded with a `try ... catch` and discard the `RecordNotFound` error.

I do think an argument could certainly be made for renaming these methods to `updateOrThrow()` and `deleteOrThrow()` and deprecating `update()` and `delete()` to be consistent with other API methods, though.

**Comment by ceri-waterscreative** [comment]
Agreed.. very odd and unexpected behaviour which isn't illustrated clearly in the documentation. 

There should be a notice/warning about this against the docs for the update statement.
https://www.prisma.io/docs/orm/prisma-client/queries/crud#update
